Rewrite of RTT estimation section (#2592)
quicwg/base-drafts: draft-ietf-quic-recovery.md

environments, this speeds recovery, reduces spurious retransmits, and ensures
forward progress without relying on timeouts.

> ### Explicit Correction For Delayed ACKs {#ack-delay}

Reviewer: Use words.  ACKs are delayed in the network, because they are frames.  Acknowledgments are the things that you put in the frame (and the things where we need to compensate for delays).

```suggestion
### Explicit Correction For Delayed Acknowledgements {#ack-delay}
```

An endpoint measures the delay incurred between when a packet is received and

> when the corresponding ACK is sent.  The endpoint encodes this host delay for

Reviewer: ```suggestion
when the corresponding acknowledgment is sent.  The endpoint encodes this host delay for
```

the largest acknowledged packet in the Ack Delay field of an ACK frame (see
Section 19.3 of {{QUIC-TRANSPORT}}).  This allows the receiver of the ACK to
adjust for any host delays - importantly, for delayed acknowledgements - when

> estimating the path RTT.  Where possible, this delay can include the time that a

Reviewer: "this delay" meaning which?  Delays introduced by the host that we want to include in the RTT, or delays introduced by the host that we **don't** want to include?

packet was held in the OS kernel before being processed by a userspace QUIC
stack.

> An endpoint MUST NOT excessively delay acknowledgements of ack-eliciting

Reviewer: I think that the most important thing here is the contract that `max_ack_delay` implies: An endpoints promises to never delay acknowledgments of an ack-eliciting packet by more than the indicated value.  If they do, any excess accrues to the RTT estimate and could result in timeouts.

Author: Heh -- I used the word contract and then removed it. I'll use your words.

### Explicit Correction For Delayed Acknowledgements {#ack-delay}

An endpoint measures the delay incurred between when a packet is received and
when the corresponding acknowledgment is sent.  The endpoint encodes this host
delay for the largest acknowledged packet in the Ack Delay field of an ACK frame
(see Section 19.3 of {{QUIC-TRANSPORT}}).  This allows the receiver of the ACK
to adjust for any host delays - importantly, for delayed acknowledgements - when
estimating the path RTT.  In certain deployments, a packet might be held in the
OS kernel or elsewhere on the host before being processed by the QUIC
stack. Where possible, an endpoint SHOULD include these delays when populating

> the Ack Delay field in an ACK frame.

Reviewer: Probably not for this change, but do we really want to encourage people to be estimating delays in the kernel?  Because it seems like the point of some of the other changes is to include that sort of delay in the RTT estimate.  What matters here is end-to-end delays - the time between when the packet was consumed and when the packet containing the ACK was sent out.  Yes, it's true that you get a bigger RTT estimate if you don't correct for delays in the kernel or NIC, but those are just another piece of the plumbing now.  If we fail to cover things like scheduling delays in the RTT estimate, we're just deluding ourselves about where the delays are.

Author: Yeah, I was thinking about this too, but it's text that we already had. FWIW, I think it's useful to separate out these delays, not because they shouldn't get used for PTO estimates, but because knowing component delays might be helpful. That said, I don't have a good use case for these component delays yet. OTOH, I've started that separation between stack-controlled delays and other delays in the text, it might be worth using that distinction here, and arguing that these delays ought to be part of the path RTT since they're not controlled delays. I'll file a separate issue.

Author: Filed #2596

An endpoint MUST NOT excessively delay acknowledgements of ack-eliciting
packets.  Specifically, an endpoint is expected to enforce a maximum delay to

> avoid causing spurious timeouts at the peer.  The maximum ack delay is

Reviewer: The text doesn't say what delay is being "enforced".  Maybe:

>[...] an endpoint is expected to send acknowledgments within its advertised delay to avoid causing spurious timeouts at its peer.

But maybe you can remove this sentence now that the "contract" text is here.

## Generating RTT samples {#latest-rtt}

An endpoint generates an RTT sample on receiving an ACK frame that meets the
following two conditions:

- the largest acknowledged packet number is newly acknowledged, and

- at least one of the newly acknowledged packets was ack-eliciting.

The RTT sample, latest_rtt, is generated as the time elapsed since the largest
acknowledged packet was sent:

~~~
latest_rtt = ack_time - send_time_of_largest_acked
~~~

An RTT sample is generated using only the largest acknowledged packet in the
received ACK frame.  This is because a peer reports host delays for only the
largest acknowledged packet in an ACK frame.  While the reported host delay is
not used by the RTT sample measurement, it is used to adjust the RTT sample in
subsequent computations of smoothed_rtt and rttvar {{smoothed-rtt}}.

To avoid generating multiple RTT samples using the same packet, an ACK frame
SHOULD NOT be used to update RTT estimates if it does not newly acknowledge the
largest acknowledged packet.

An RTT sample MUST NOT be generated on receiving an ACK frame that does not
newly acknowledge at least one ack-eliciting packet.  A peer does not send an

> ACK frame on receiving only non-ack-eliciting packets, and an ACK frame that is

Reviewer 2: ```suggestion
ACK frame on receiving only non-ack-eliciting packets, so an ACK frame that is
```

loss recovery algorithm may cause spurious retransmits, but the sender will
continue making forward progress.

## Measuring and Reporting Host Delay {#host-delay}

> An endpoint measures the delay incurred between when a packet is received and

Reviewer 2: Ping

> An endpoint measures the delay incurred between when a packet is received and

Author: See my response above.

when the corresponding acknowledgment is sent.  The endpoint encodes this host
delay for the largest acknowledged packet in the Ack Delay field of an ACK frame
(see Section 19.3 of {{QUIC-TRANSPORT}}).  This allows the receiver of the ACK

> to adjust for any host delays - importantly, for delayed acknowledgements - when

Reviewer 2: Bump?

> to adjust for any host delays - importantly, for delayed acknowledgements - when

Reviewer 2: Ping

> to adjust for any host delays - importantly, for delayed acknowledgements - when

Author: I'm not sure I follow. It's fine to use commas and dashes together. Can you suggest a rephrasing?

> to adjust for any host delays - importantly, for delayed acknowledgements - when

Reviewer 2: ```suggestion
to adjust for any host delays, which is particularly important for delayed acknowledgements, when
```

estimating the path RTT.  In certain deployments, a packet might be held in the
OS kernel or elsewhere on the host before being processed by the QUIC
stack. Where possible, an endpoint SHOULD include these delays when populating
the Ack Delay field in an ACK frame.

An endpoint MUST NOT excessively delay acknowledgements of ack-eliciting
packets.  The maximum ack delay is communicated in the max_ack_delay transport

> parameter, see Section 18.1 of {{QUIC-TRANSPORT}}.  max_ack_delay implies an

Author: I think implies works better here

> parameter, see Section 18.1 of {{QUIC-TRANSPORT}}.  max_ack_delay implies an

Reviewer 2: I would disagree, since I do not like the word implies in IETF documents.

> parameter, see Section 18.1 of {{QUIC-TRANSPORT}}.  max_ack_delay implies an

Author: https://www.merriam-webster.com/dictionary/imply --> "imply" is commonly used to indicate necessary consequence, not just inference (I suspect your dislike is to this meaning of imply).

The way I read it, max_ack_delay specifies a transport parameter value which implies the contract.

> parameter, see Section 18.1 of {{QUIC-TRANSPORT}}.  max_ack_delay implies an

Reviewer 2: Yeah, not in love with it here, but I'm failing to convince you, so let's leave it as is.

explicit contract: an endpoint promises to never delay acknowledgments of an
ack-eliciting packet by more than the indicated value. If it does, any excess
accrues to the RTT estimate and could result in spurious retransmissions from
the peer.

# Estimating the Round-Trip Time {#compute-rtt}

> At a high level, an endpoint measures the time from when a packet was sent to

Author: It's not quite accurate to say that this is an RTT sample because of all the conditions below, which is why I have the phrase.

when it is acknowledged as a round-trip time (RTT) sample.  The endpoint uses
RTT samples and peer-reported host delays ({{host-delay}}) to generate a
statistical description of the connection's RTT.  An endpoint computes the
following three values: the minimum value observed over the lifetime of the
connection (min_rtt), an exponentially-weighted moving average (smoothed_rtt),
and the variance in the observed RTT samples (rttvar).

## Generating RTT samples {#latest-rtt}

An endpoint generates an RTT sample on receiving an ACK frame that meets the
following two conditions:

- the largest acknowledged packet number is higher than any previously seen, and

- at least one of the newly acknowledged packets was ack-eliciting.

The RTT sample, latest_rtt, is generated as the time elapsed since the largest
acknowledged packet was sent:

~~~
latest_rtt = ack_time - send_time_of_largest_acked
~~~

> An endpoint uses only locally observed times in generating RTT samples and does

Author: This subsection is talking about generating RTT samples, which do not adjust for host delays.

> An endpoint uses only locally observed times in generating RTT samples and does

Reviewer 2: I would just remove this paragraph, because it's potentially confusing and I'm not sure what it's adding?

> An endpoint uses only locally observed times in generating RTT samples and does

Author: Fair, it's not necessary yet, so removed.

not adjust for any host delays reported by the peer ({{host-delay}}).

A peer reports host delays for only the largest acknowledged packet in an ACK

> frame, which is assumed by subsequent computations of smoothed_rtt and rttvar in

Author: Same as above -- this section on generating RTT samples doesn't use host delay, but later subsections (srtt, rttvar) do.

> frame, which is assumed by subsequent computations of smoothed_rtt and rttvar in

Reviewer 2: Bump, I'm still unclear on what the latter half of the sentence is trying to communicate.

> frame, which is assumed by subsequent computations of smoothed_rtt and rttvar in

Author: I've rewritten the para, so hopefully it reads better.

adjusting for host delays.  As a result, an RTT sample is only generated using
the largest acknowledged packet in the received ACK frame.

To avoid generating multiple RTT samples using the same packet, an ACK frame

> SHOULD NOT be used to update RTT estimates if it does not newly acknowledge the

Author: good call -- fixed the condition above.

largest acknowledged packet.

An RTT sample MUST NOT be generated on receiving an ACK frame that does not

> newly acknowledge at least one ack-eliciting packet.  A peer does not send an

Author: We are clear on the MUST NOT, and I don't want to dwell too long on the rationale... happy to take any text suggestions here, but I really don't think it's worth getting into the complications.

> newly acknowledge at least one ack-eliciting packet.  A peer does not send an

Reviewer 2: Suggestion below

loss recovery algorithm may cause spurious retransmits, but the sender will
continue making forward progress.

> # Computing the round-trip time (RTT) estimate {#compute-rtt}

Reviewer: Use Title Case.  And drop the acronym from the title.

```suggestion
# Estimating the Round-Trip Time {#compute-rtt}
```

At a high level, an endpoint measures the time from when a packet was sent to
when it is acknowledged as a round-trip time (RTT) sample.  The endpoint uses
RTT samples and peer-reported host delays ({{ack-delay}}, {{generating-acks}}) to
generate a statistical description of the connection's RTT.  An endpoint
computes the following three values: the minimum value observed over the
lifetime of the connection (min_rtt), an exponentially-weighted moving average
(smoothed_rtt), and the variance in the observed RTT samples (rttvar).

## Generating RTT samples {#latest-rtt}

An endpoint generates an RTT sample on receiving an ACK frame that meets the
following two conditions:

- the largest acknowledged packet number is higher than any previously seen, and

- at least one of the newly acknowledged packets was ack-eliciting.

The RTT sample, latest_rtt, is generated as the time elapsed since the largest
acknowledged packet was sent:

~~~
latest_rtt = ack_time - send_time_of_largest_acked
~~~

An endpoint uses only locally observed times in generating RTT samples and does
not correct for any host delays reported by the peer ({{ack-delay}}).

> An RTT sample MUST NOT be generated for a packet that is not newly acknowledged.

Reviewer: Do we generate samples for packets?

```suggestion
An ACK frame that does not newly acknowledge any packets MUST NOT be used to update RTT estimates.
```

This avoids multiple RTT samples for the same packet.

An RTT sample MUST NOT be generated for a packet that is not the largest

> acknowledged in the received ACK frame.  A peer reports host delays for only the

Reviewer: I don't think that you need a "MUST NOT" here.  Your definition of the sample covers this.

largest acknowledged packet in an ACK frame, which is assumed by subsequent
computations of smoothed_rtt and rttvar in correcting for host delays.

An RTT sample MUST NOT be generated if the ACK frame does not newly acknowledge
an ack-eliciting packet. A peer does not send an ACK frame in response to
receiving only non-ack-eliciting packets, and an ACK frame that is subsequently
sent can include an arbitrarily large Ack Delay field.  Ignoring such ACK frames
avoids complications in subsequent smoothed_rtt and rttvar computations.

A sender might generate multiple RTT samples per RTT when multiple ACK frames
are received within an RTT.  As suggested in {{?RFC6298}}, doing so might result

> in inadequate history in smoothed_rtt and rttvar.  This is an open question

Reviewer: The "This" here is not very clear.  "Ensuring that RTT estimates retain sufficient history is an open research question."

which we leave for future research to address.

## Computing min_rtt {#min-rtt}

min_rtt is the minimum RTT observed over the lifetime of the connection.
min_rtt is set to the latest_rtt on the first sample in a connection, and to the
lesser of min_rtt and latest_rtt on subsequent samples.

An endpoint uses only locally observed times in computing the min_rtt and does
not correct for host delays reported by the peer ({{ack-delay}}).  Doing so
allows the endpoint to set a lower bound for the smoothed_rtt based entirely on
what it observes (see {{smoothed-rtt}}), and limits potential underestimation
due to erroneously-reported delays by the peer.

## Computing smoothed_rtt and rttvar {#smoothed-rtt}

smoothed_rtt is an exponentially-weighted moving average of an endpoint's
RTT samples, and rttvar is the endpoint's estimated variance in
the RTT samples.

smoothed_rtt estimates path latency by correcting RTT samples for peer-reported

> host delays ({{ack-delay}}).  A peer indicates that the peer's delay in sending

Reviewer: Check grammar.  "A peer indicates that the peer's delay" might be correct, but it's hard to read.

an acknowledgement for an ack-eliciting packet will be no greater than the
max_ack_delay transport parameter.  Consequently, when a peer reports an Ack
Delay that is greater than max_ack_delay, the delay is attributed to reasons out
of the peer's control, such as scheduler latency at the peer or loss of previous
ACK frames.  Any delays beyond the peer's max_ack_delay are therefore considered
effectively part of path delay and incorporated into the smoothed_rtt estimate.

When correcting an RTT sample using peer-reported acknowledgement delays, an
endpoint:

- SHOULD use the lesser of the value reported in Ack Delay field of the ACK

>   frame and the peer's max_ack_delay transport parameter ({{ack-delay}}).

Reviewer: Why SHOULD?  Why use 2119 language here at all?

> When correcting an RTT sample, endpoints first limit the reported Ack Delay to the value of the max_ack_delay transport parameter.  The RTT sample is then reduced by the resulting delay value.  If correcting for host delays would result in a sample that is less than min_rtt, the unmodified sample is used.  This guards against underestimation of RTT as a result of inflated reporting of host delays.

Author: Fair point.  Ultimately, this is bounded by another value that the peer advertises (max_ack_delay), so there's no additional safety issue here.

- MUST not apply the correction if the resulting RTT sample is smaller than the
  min_rtt.  This limits the underestimation that a misreporting peer can cause
  to the smoothed_rtt.

On the first RTT sample in a connection, the smoothed_rtt is set to the
latest_rtt.  The peer-reported host delay MUST NOT be used on the first sample,
since correcting for it would result in a smoothed_rtt that is smaller than the

> min_rtt.

Reviewer: This assertion isn't necessary.  Yes, the one sample you have for min_rtt is equal to latest_rtt, so of course any host delay will not be applied.  So this just duplicates the earlier requirement.  If anything, just note that there is no point in applying host delay corrections as they are never applied.

Author: I think the way the text flows now makes this statement moot. I'm removing it.

smoothed_rtt and rttvar are computed as follows, similar to {{?RFC6298}}.  On
the first RTT sample in a connection:

~~~
smoothed_rtt = latest_rtt
rttvar = latest_rtt / 2
~~~

On subsequent RTT samples, smoothed_rtt and rttvar evolve as follows:

~~~
ack_delay = min(Ack Delay in ACK Frame, max_ack_delay)
if (min_rtt + ack_delay < latest_rtt):

>   latest_rtt = latest_rtt - ack_delay

Reviewer: I think that you should use a different name for this result.  Otherwise, you are change the value of a variable that you rely on in prose.

Author: Good point. Changed.

~~~
ack_delay = min(Ack Delay in ACK Frame, max_ack_delay)
corrected_rtt = latest_rtt
if (min_rtt + ack_delay < latest_rtt):

>   corrected_rtt = latest_rtt - ack_delay

Reviewer: Maybe `adjusted_rtt` instead.  "Corrected" implies that there is some notion of "correctness" that applies.

  // Limit ack_delay by max_ack_delay
  ack_delay = min(ack_delay, max_ack_delay)
  // Adjust for ack delay if it's plausible.
  adjusted_rtt = latest_rtt
  if (latest_rtt - min_rtt > ack_delay):
    adjusted_rtt = latest_rtt - ack_delay
  // Based on {{?RFC6298}}.
  if (smoothed_rtt == 0):

>     smoothed_rtt = adjusted_rtt

Reviewer 2: I'd suggest moving the adjusted_rtt calculation into the else and using latest_rtt here?  If not, I'd ensure smoothed_rtt and rttvar are set as a function of the same field, not one based on adjusted_rtt and the other latest_rtt.

Author: *facepalm* thanks for catching this. I believe this was a bug in the original pseudo-code and the ack_delay adjustment should have been in the else to begin with.
